File: nodecore-cli/src/main/java/nodecore/cli/serialization/CoinbaseTransactionInfo.java
// VeriBlock NodeCore CLI
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.cli.serialization;

import com.google.gson.annotations.SerializedName;
import nodecore.api.grpc.VeriBlockMessages;
import nodecore.api.grpc.utilities.ByteStringUtility;
import org.veriblock.core.utilities.Utility;

import java.util.ArrayList;
import java.util.List;

public class CoinbaseTransactionInfo {
    public CoinbaseTransactionInfo(final VeriBlockMessages.CoinbaseTransaction coinbaseTransaction)
    {
        coinbase_tx_hash = ByteStringUtility.byteStringToHex(coinbaseTransaction.getTxId());
        powCoinbaseAmount = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.getPowCoinbaseAmount());
        popCoinbaseAmount = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.getPopCoinbaseAmount());
        powFeeShare = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.getPowFeeShare());
        popFeeShare = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.getPopFeeShare());

        for (VeriBlockMessages.Output powOutput : coinbaseTransaction.getPowOutputsList()) {
            powOutputs.add(new OutputInfo(powOutput));
        }

        for (VeriBlockMessages.Output popOutput : coinbaseTransaction.getPopOutputsList()) {
            popOutputs.add(new OutputInfo(popOutput));
        }
    }

    @SerializedName("pow_coinbase_amount")
    public String powCoinbaseAmount;

    @SerializedName("pop_coinbase_amount")
    public String popCoinbaseAmount;

    @SerializedName("pow_fee_share")
    public String powFeeShare;

    @SerializedName("pop_fee_share")
    public String popFeeShare;

    @SerializedName("pow_outputs")
    public List<OutputInfo> powOutputs = new ArrayList<>();

    @SerializedName("pop_outputs")
    public List<OutputInfo> popOutputs = new ArrayList<>();

    public String coinbase_tx_hash;
}

File: nodecore-cli/src/main/java/nodecore/cli/serialization/CoinbaseTransactionInfo.kt
// VeriBlock NodeCore CLI
// Copyright 2017-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.cli.serialization

import com.google.gson.annotations.SerializedName
import nodecore.api.grpc.RpcCoinbaseTransaction
import nodecore.api.grpc.utilities.extensions.toHex
import org.veriblock.core.utilities.Utility

class CoinbaseTransactionInfo(
    coinbaseTransaction: RpcCoinbaseTransaction
) {
    @SerializedName("pow_coinbase_amount")
    val powCoinbaseAmount = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.powCoinbaseAmount)

    @SerializedName("pop_coinbase_amount")
    val popCoinbaseAmount = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.popCoinbaseAmount)

    @SerializedName("pow_fee_share")
    val powFeeShare = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.powFeeShare)

    @SerializedName("pop_fee_share")
    val popFeeShare = Utility.formatAtomicLongWithDecimal(coinbaseTransaction.popFeeShare)

    @SerializedName("pow_outputs")
    val powOutputs = coinbaseTransaction.powOutputsList.map { output ->
        OutputInfo(output)
    }

    @SerializedName("pop_outputs")
    val popOutputs = coinbaseTransaction.popOutputsList.map { output ->
        OutputInfo(output)
    }

    val coinbase_tx_hash = coinbaseTransaction.txId.toHex()
}
